### ISSUE TITLE
[BUG] Cannot get elastic scaling to work

### ISSUE THREAD

**Issue by marwage** [microsoft/DeepSpeed#3146]
**Cannot elastically scale up the number of workers**
When I want to scale up the number of workers from 2 to 3, the training crashes.

**To Reproduce**
1. Use https://github.com/microsoft/DeepSpeedExamples/blob/arpan/elastic_scripts/elastic_training/train_bert_ds_el_script.py
2. Run
```
deepspeed --num_nodes 2 --max_num_nodes 4 --min_num_nodes 1 \
          --num_gpus 1 --force_multi --hostfile hosts_el \
          --elastic_training --master_port 49091 --master_addr worker-0  \
          train_bert_ds_el_script.py --checkpoint_dir ./checks_el
```
3. Run the same command with hostfile that has 3 workers on worker with rank 3.

**Expected behavior**
Continue training with 3 workers.

**Error**
```
deepspeed-worker-2-1  | worker-0: [W socket.cpp:426] [c10d] The server socket has failed to bind to [::]:49091 (errno: 98 - Address already in use).
deepspeed-worker-2-1  | worker-0: [W socket.cpp:426] [c10d] The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).
deepspeed-worker-2-1  | worker-0: [E socket.cpp:462] [c10d] The server socket has failed to listen on any local network address.
```
```
deepspeed-worker-0-1  | worker-0: Sending process 175 closing signal SIGTERM
deepspeed-worker-0-1  | worker-1: Sending process 100 closing signal SIGTERM
deepspeed-worker-2-1  | worker-1: => set cuda device = 0 Global Rank = 3
deepspeed-worker-0-1  | worker-1: => set cuda device = 0 Global Rank = 2
deepspeed-worker-2-1  | worker-0: => set cuda device = 0 Global Rank = 1
deepspeed-worker-0-1  | worker-0: => set cuda device = 0 Global Rank = 0
deepspeed-worker-2-1  | worker-2: => set cuda device = 0 Global Rank = 4
```
```
deepspeed-worker-2-1  | worker-0: Duplicate GPU detected : rank 1 and rank 0 both on CUDA device 1000
```

**System info (please complete the following information):**
 - OS: Ubuntu 22.04
 -  1 A6000 GPU per worker
 - Interconnects Ethernet
 - Python version 3.10

**Launcher context**
`deepspeed` with `--elastic_training`

**Docker context**
Image based on nvidia/cuda:11.7.1-cudnn8-devel-ubuntu22.04
With Deepspeed from https://github.com/microsoft/DeepSpeed/tree/arpan/elasticity 

**Assumption**
The new worker is not added to the training but an now training with 3 workers is started.

**Comment by awan-10** [comment]
@aj-prime -- can you please take a look at this?

**Comment by awan-10** [comment]
@marwage -- thanks for trying out this feature. It has not been merged into our main branch. If Arpan (@aj-prime) and you can make this work, we can plan for that.

**Comment by aj-prime** [comment]
Sure, @awan-10. I can take a look. 

@marwage - How many GPUs do you have per node? 

It looks like two workers are running on same GPU.  

`deepspeed-worker-2-1  | worker-0: Duplicate GPU detected : rank 1 and rank 0 both on CUDA device 1000`

**Comment by marwage** [comment]
Hi @awan-10, hi @aj-prime,
Thanks for getting back to me!

I am using 1 host with 4 containers which have 1 GPU each. I start the training with 2 containers and then add the additional 2 containers. The training starts with the two containers.

```
deepspeed-worker-2-1  | worker-0: [W socket.cpp:426] [c10d] The server socket has failed to bind to [::]:49091 (errno: 98 - Address already in use).
deepspeed-worker-2-1  | worker-0: [W socket.cpp:426] [c10d] The server socket has failed to bind to ?UNKNOWN? (errno: 98 - Address already in use).
```
That looks like that when I run the `deepseed` command on a new worker, wants to run a new rendezvous server.

```
deepspeed-worker-2-1  | worker-0: => set cuda device = 0 Global Rank = 1
deepspeed-worker-2-1  | worker-2: => set cuda device = 0 Global Rank = 4
deepspeed-worker-2-1  | worker-1: => set cuda device = 0 Global Rank = 3
deepspeed-worker-0-1  | worker-0: 2023-05-25 15:09:30.814 | INFO     | __main__:log_dist:58 - [Rank 0] Starting Script
deepspeed-worker-0-1  | worker-0: => set cuda device = 0 Global Rank = 0
deepspeed-worker-0-1  | worker-1: => set cuda device = 0 Global Rank = 2
deepspeed-worker-2-1  | worker-3: => set cuda device = 0 Global Rank = 5
```
While this looks like it adds worker 0 to 4 in addition to the already running worker 0 and 1.

I am using `docker compose` to run the training and you can find all my code here https://github.com/marwage/ds-elastic-exper

**Comment by aj-prime** [comment]
Yeah, it is adding more ranks than expected. How are you creating a container? 
Elastic support is at node-level; therefore, I think you are seeing this error. Can you test it node-level?

**Comment by marwage** [comment]
Ok, thanks, I was running it on the VM level and got the same error.
Turns out that `--hostfile` should only contain the new worker and I gave it the full list of workers. Second, the the `deespeed` command must be run on all the new workers -- not only on one of the new workers. It would be nice to include that in your doc.

I have a follow up question regarding loading the checkpoint. In the code you are checking for the checkpoint path
https://github.com/microsoft/DeepSpeedExamples/blob/06b6a24c4e666f04cf97ecc692f1990993876a2d/elastic_training/train_bert_ds_el_script.py#L820C5-L820C5
But on the new workers there will be no checkpoint. How do you solve that problem?
